spring-kafka/src/main/java/org/springframework/kafka/listener/DefaultAfterRollbackProcessor.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.listener;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.TopicPartition;
+
+/**
+ * Default implementation of {@link AfterRollbackProcessor}. Seeks all
+ * topic/partitions so the records will be re-fetched, including the failed
+ * record.
+ *
+ * @param <K> the key type.
+ * @param <V> the value type.
+ *
+ * @author Gary Russell
+ *
+ * @since 1.3.5
+ *
+ */
+public class DefaultAfterRollbackProcessor<K, V> implements AfterRollbackProcessor<K, V> {
+
+	private static final Log logger = LogFactory.getLog(DefaultAfterRollbackProcessor.class);
+
+	@Override
+	public void process(List<ConsumerRecord<K, V>> records, Consumer<K, V> consumer) {
+		Map<TopicPartition, Long> seekOffsets = new HashMap<>();
+		Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
+		while (iterator.hasNext()) {
+			ConsumerRecord<K, V> record = iterator.next();
+			TopicPartition topicPartition = new TopicPartition(record.topic(), record.partition());
+			if (!seekOffsets.containsKey(topicPartition)) {
+				seekOffsets.put(topicPartition, record.offset());
+			}
+		}
+		for (Entry<TopicPartition, Long> entry : seekOffsets.entrySet()) {
```

Comment:
Why do we need this extra loop if we can just perform `seek` during the first one skipping already processed partitions ?

Reply:
It seems much more complicated to me to keep track of which ones we've already seeked.

Reply:
I don't see a big difference and I doubt the performance will be any better - it has to calculate hash codes instead performing the second loop - but feel free to change it if you insist.

Reply:
>  it has to calculate hash codes 

???

Don't you do that already with the current `HashMap` and `seekOffsets.containsKey()` ?

Reply:
Doh, of course; ok. Feel free to change it.